core/src/kura.rs

```diff
@@ -4,258 +4,475 @@
+/// in a "Block Store". This trait is the interface to such a
+/// block store. It implements certain functionality for kura to
+/// use. And it defines what functionality needs to be implemented
+/// for every platform. The default implementation is `StdFileBlockStore`
+/// which uses rust's `std::fs`.
+pub trait BlockStoreTrait {
+    /// Read a series of block indices from the block index file.
+    /// Attempt to fill all of `dest_buffer`.
+    /// # Errors
+    /// IO Error.
+    fn read_block_indices(
+        &self,
+        start_block_height: u64,
+        dest_buffer: &mut [(u64, u64)],
+    ) -> Result<()>;
+    /// Write a single block index at the specified `block_height`.
+    /// If `block_height` is beyond the end of the index file attempt to
+    /// extend the index file.
+    /// # Errors
+    /// IO Error.
+    fn write_block_index(&mut self, block_height: u64, start: u64, length: u64) -> Result<()>;
 
-    async fn handle(&mut self, StoreBlock(block): StoreBlock) {
-        #[cfg(feature = "telemetry")]
-        if block.header().height == 1 {
-            iroha_logger::telemetry!(msg = iroha_telemetry::msg::SYSTEM_CONNECTED, genesis_hash = %block.hash());
-        }
-        if let Err(error) = self.store(block).await {
-            error!(%error, "Failed to write block")
+    /// Read the amount of indices in the index file. (Byte size of
+    /// the index file divided by 16 B.)
+    /// # Errors
+    /// IO Error. Because this function
+    /// does so little, if this fails you can be quite sure all other
+    /// read and write ops will fail. The most common reason this function
+    /// errors is because you forgot to call `create_files_if_they_do_not_exist`.
+    fn read_index_count(&self) -> Result<u64>;
+    /// Change the value returned by `read_index_count` aka the size of the
+    /// index file.
+    /// # Errors
+    /// IO Error.
+    fn write_index_count(&mut self, new_count: u64) -> Result<()>;
+
+    /// Read from the data file in order to fill `dest_buffer` starting
+    /// from `start_location_in_data_file`.
+    /// # Errors
+    /// IO Error.
+    fn read_block_data(
+        &self,
+        start_location_in_data_file: u64,
+        dest_buffer: &mut [u8],
+    ) -> Result<()>;
+    /// Write `block_data` into the data file starting at
+    /// `start_location_in_data_file`. Extend the file if
+    /// necessary.
+    /// # Errors
+    /// IO Error.
+    fn write_block_data(
+        &mut self,
+        start_location_in_data_file: u64,
+        block_data: &[u8],
+    ) -> Result<()>;
+
+    /// Create the index and data files if they do not
+    /// already exist.
+    /// # Errors
+    /// Will fail if the any of the files don't exist
+    /// and couldn't be created.
+    fn create_files_if_they_do_not_exist(&mut self) -> Result<()>;
+
+    /// Above are the platform dependent functions.
+    /// Below are the platform independent functions that use the above functions.
```

Comment:
I think these two lines should be non-doc comments (`//`). 

I'd suggest adding a doc comment to each function saying whether or not it is platform independent

Reply:
good idea

Reply:
Well okay, they are all technically platform dependent from the "users" perspective. But those two comments should not be doc comments.

Reply:
They are comments for the programmers in the future implementing a block store.